### PR TITLE
Fix: Enable Recent Activity Table Item Redo Functionality

### LINKED ISSUE #3551: QA Dashboard: Redo action trigger from activity table
## Steps to reproduce

1. Ensure there are historic actions within the activity table on the dashboard.
2. Using the meatball menu (...) select the re-do action from the dropdown list.
3. Nothing happens and the action panel of the same action type is not triggered and relevant info is not pre-populated. 

image

## Expected behaviour

* The redo menu item should work correctly, opening a new action panel once selected and populating the same data as the transaction it was selected from. It should work across all responsive sizes.

## Actual behaviour

* Nothing happens when the redo action menu item is selected from within the dropdown.

### LINKED ISSUE #3593: [Actions - QA ]App Crashes on redo of "Manage Reputation" due to invalid BigNumber Error
## Steps to reproduce

1. Open Manage reputation action draw.
2. Attempt to redo actions.
3. Observe the app behavior.

## Expected behaviour

* Redo "Manage reputation" action should proceed smoothly without causing any app crash.

## Actual behaviour

* The app crashes, and the console displays an "invalid BigNumber" error

image

image

### PR DESCRIPTION
**Pull request by rumzledz**
## Description

This PR enables the Dashboard page's Activity Table Item Redo Function and aligns it with Activity page's Activity Table.

Redo action

## Testing

> [!NOTE]
> I had to fix the following existing `master` bug and do some refactoring as well. It is unrelated to issue #3551 but you know what, it makes sense to include it in this PR:
>
> @iamsamgibbs  "The only reason I'm rejecting is that not all actions should have the redo option"
> @rdig  "I've tested this with Simple Payments and which work as expected, however, as Sam noted, not all actions require it, so this needs to be enabled on a case by case basis"
>
> I've added a testing section for it 😄 
>
> I also had to handle the Redo flow for Reputation-related actions in the Activity Table items so please test that as well.
>
> I'll find a way to refactor the Completed Actions component as well, so we'll have a single source of truth for knowing which actions are not allowed to have a Redo Action. But it will be in a separate future PR if there's capacity 👌 

### 1. Verify that the Redo action button does what it's intended to do

1. Go to the Dashboard
2. Make a simple payment
3. Once it shows up on the Recent Activity table, click its kebab button
4. Click Redo action
5. Verify that the Action Form comes up and the details are filled in
6. Visit the Activity page http://localhost:9091/planex/activity
7. Look for the same action on the table and redo it
8. Verify that the result is consistent as step 4
 
### 2. Verify that the Redo action is conditionally rendered on a case-by-case basis

1. Go to the Dashboard
2. Create a new Team
3. Add the following cases in `useBuildRedoEnabledActionsMap.ts`, after line 116:

```js
case ColonyActionType.MoveFunds:
```
image

4. Reload the Dashboard
5. Click the kebab menu for the Create Domain action
6. Verify that you don't see Redo action in the popup menu

image

7. Click the kebab menu for any Payment action
9. Verify that the Redo action is still present
> [!NOTE]
> The pagination might reset back to page 1 but this is already being fixed by @mmioana in PR: https://github.com/JoinColony/colonyCDapp/pull/3566
10. Move through the Activity table's pages until you find the Transfer Funds action which is done as part of our `create-data` script and click its kebab menu
11. Verify that you don't see Redo action in the popup menu
12. Please repeat steps 1 - 10 for the Actions Table in the Activity page: http://localhost:9091/{colonyName}/activity

### 3. Verify that you can redo Reputation actions and that the app doesn't crash in the process

1. Award 20,000 points for Amy
2. Redo the action via the UserHub Transactions tab
3. Verify that the app does not crash and Manage Reputation form shows up
4. Redo the action via the Activity table
5. Verify that the app does not crash and Manage Reputation form shows up

## Diffs

**New stuff** ✨

* I centralised this "Redo action" behaviour for the Actions Table items in a new hook called `useRedoAction` so it's shared consistently between the Action Tables found with the Dashboard & Activity pages.

Resolves #3551
Resolves #3593

### CODE REVIEW

File: src/components/common/ColonyActionsTable/hooks/useRedoAction.ts
Comment (mmioana) [inline src/components/common/ColonyActionsTable/hooks/useRedoAction.ts:1]: Really nice 🙌

**Review by mmioana — DISMISSED** [review]
Nice, now we're able to redo an action from both the `Recent activity` table on `Dashboard` and from the table on the `Activity` page @rumzledz 🎉 

https://github.com/user-attachments/assets/211d2639-0899-4267-81d9-8a0e39a93650


https://github.com/user-attachments/assets/d0011bfd-1637-4eac-86b8-c432c3c34101

**Comment by rumzledz** [comment]
Hey @rdig @iamsamgibbs all right noted. This seems to be a wider bug since the Colony Actions have had this Redo Action for all actions for the past 2 months and no one noticed. I'll see what actions are allowed to have this Redo button, will amend the logic for both the Dashboard and Activity page's actions table and will update the issue's dev hours field 👌

**Comment by rumzledz** [comment]
Hey @iamsamgibbs @rdig I've pushed the required changes. Catch you both on Monday! 👋

**Review by iamsamgibbs — DISMISSED** [review]
Great job with this!

Screenshot 2024-11-04 at 10 31 17

Screenshot 2024-11-04 at 10 31 24

Nice implementation. I agree with your comments that it would be nice to rework the completed action component so that there is a single source of truth for this. I was actually thinking earlier about how spread out everything to do with actions is, and that we could either benefit with unifying aspects of it (I believe the work Chris is doing will help with this), or at the very least, the next time we add a new action we could document everywhere new code is needed as it is so spread out.

Thanks for adding the new functionality around when to show / not show the redo action, appreciate it was outside of the original scope, but you've got it working very nicely.

**Review by bassgeta — DISMISSED** [review]
Really nice solution with the map, it's much nicer and much more unified this way
Create a new team doesn't have Redo :heavy_check_mark: 
image
Transfer funds works :heavy_check_mark: 
image
Simple payment :heavy_check_mark: 
image
image
Permissions also work :heavy_check_mark: 
image
image
Activity feed also works :100: 
Nice work with solving this, it's quite a simple hook which does one thing only, super legit :muscle:

**Comment by rumzledz** [comment]
Hey @bassgeta @iamsamgibbs! @rdig has just helped me uncover that the Activity table cannot currently redo Reputation-related actions. So I had to push that which got rid of your reviews 😅 

The latest commit also includes a fix for preventing the app to crash when redoing Reputation-related actions, which is an issue raised in #3593 

I thought it made sense to combine these fixes because they intertwine.

I will be updating the test steps shortly so you can verify that the app doesn't crash when redoing a Reputation action 👌

**Review by rdig — DISMISSED** [review]

**Review by iamsamgibbs — DISMISSED** [review]
Nice job! Awarding reputation can be redone:

https://github.com/user-attachments/assets/cf402d90-376f-4711-bbcb-75e307e26215

Redo option is being conditionally rendered as expected:

Screenshot 2024-11-06 at 09 52 21

**Review by rdig — APPROVED** [review]
Nice turnaround of this. Tested and everything works as expected.

Also, nice job in centralizing the redo actions.

Redo simple payment

https://github.com/user-attachments/assets/b68f7726-b50c-4a48-b277-c3c8f6865397

No redo on a "Create Team" action

https://github.com/user-attachments/assets/3993ab80-655c-4802-b188-2c86e17e660f

Award reputation redo does not crash the app

https://github.com/user-attachments/assets/9f6558d5-4520-4b7a-ab04-4009e7152c54

**Comment by rumzledz** [comment]
@iamsamgibbs @rdig @iamsamgibbs apologies guys, I had to resolve some merge conflicts in `SetUserRoles.tsx`, the reviews got wiped 😅

**Review by rdig — APPROVED** [review]
Reapproved

**Review by bassgeta — APPROVED** [review]
Reapproving :+1:

**Review by iamsamgibbs — COMMENTED** [review]
I just went to test this and make sure redoing Manage Permissions was still working, but it's suffering from the action form not working issue. Can you rebase it @rumzledz ?

**Comment by rumzledz** [comment]
> I just went to test this and make sure redoing Manage Permissions was still working, but it's suffering from the action form not working issue. Can you rebase it @rumzledz ?

Ah right of course, new fixes were added to `master` this morning 🥲 Will rebase!

**Comment by rumzledz** [comment]
@iamsamgibbs rebased! 🚀

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Really minor issue, but when redoing a manage permissions action, the authority type is not properly set. Here is an example where it should be multi-sig:

https://github.com/user-attachments/assets/363fac58-e52a-4e87-ba7d-a767d7c3c6ac

Works fine when redoing from the completed action, but not the activity table.

I also noticed this fun behaviour. Seems to only happen on the manage permissions form:

https://github.com/user-attachments/assets/529c06a4-a9fd-4362-9742-fee2809d23a9

And this is happening when redoing via either method:

https://github.com/user-attachments/assets/91ade709-ea73-4a67-b462-c661cf2947a5

Could always open this as a new issue for the sake of getting it merged. Not sure if any of these are specifically introduced by this PR.

**Review by iamsamgibbs — APPROVED** [review]
Approving so we can get this merged and we can open a new issue for the last little bits

**Pull request merged**